src/ring_theory/graded_algebra/basic.lean

```diff
@@ -130,30 +173,39 @@ end graded_algebra
 
 section canonical_order
 
-open graded_algebra set_like.graded_monoid direct_sum
+open graded_ring set_like.graded_monoid direct_sum
 
-variables {ι R A : Type*}
-variables [comm_semiring R] [semiring A]
-variables [algebra R A] [decidable_eq ι]
+variables [semiring A] [decidable_eq ι]
 variables [canonically_ordered_add_monoid ι]
-variables (𝒜 : ι → submodule R A) [graded_algebra 𝒜]
+variables [set_like σ A] [add_submonoid_class σ A] (𝒜 : ι → σ) [graded_ring 𝒜]
 
 /--
 If `A` is graded by a canonically ordered add monoid, then the projection map `x ↦ x₀` is a ring
 homomorphism.
 -/
 @[simps]
-def graded_algebra.proj_zero_ring_hom : A →+* A :=
+def graded_ring.proj_zero_ring_hom : A →+* A :=
 { to_fun := λ a, decompose 𝒜 a 0,
   map_one' := decompose_of_mem_same 𝒜 one_mem,
-  map_zero' := by simp only [subtype.ext_iff_val, decompose_zero, zero_apply, submodule.coe_zero],
-  map_add' := λ _ _, by simp [subtype.ext_iff_val, decompose_add, add_apply, submodule.coe_add],
-  map_mul' := λ x y,
-    have m : ∀ x, x ∈ supr 𝒜,
-    from λ x, (graded_algebra.is_internal 𝒜).submodule_supr_eq_top.symm ▸ submodule.mem_top,
-    begin
-    refine submodule.supr_induction 𝒜 (m x) (λ i c hc, _) _ _,
-    { refine submodule.supr_induction 𝒜 (m y) (λ j c' hc', _) _ _,
+  map_zero' := by simp,
+  map_add' := λ _ _, by simp,
+  map_mul' := λ x y, begin
+    -- Convert the abstract add_submonoid into a concrete one. This is necessary as there is no
+    -- lattice structure on the abstract ones.
+    let 𝒜' : ι → add_submonoid A :=
+      λ i, (⟨𝒜 i, λ _ _, add_mem_class.add_mem, zero_mem_class.zero_mem _⟩ : add_submonoid A),
```

Comment:
This suggests there should be an instance `add_submonoid_class S M → has_coe S (add_submonoid M)` (probably `has_coe_t` now that I think of it), so we can just use `coe` here.

Reply:
I agree, although there's probably a whole family of such coercions (and associated simp lemmas), and I didn't want the scope to creep too much here.

Reply:
OK, let's leave it for a follow-up. I've created #15053 to track it.